File: web/src/grpc/mux.ts
// One socket for every live stream.
//
// gRPC-web runs each server-stream on its own HTTP request, and a browser
// allows six connections PER HOST across every tab. The editor holds three
// long-lived streams per page (project or flow, statistics, activity), so two
// open tabs exhausted the pool exactly and every request afterwards — the one
// carrying widget data included — queued forever. The dashboard rendered
// empty cards and the page looked frozen.
//
// A WebSocket is one connection and is not subject to that cap. This client
// carries every stream over a single socket to tiny's /ws endpoint, addressed
// by id. The server sends the same protobuf messages its gRPC handlers
// already produce, so each event is decoded with the generated message class
// and the editor receives exactly what gRPC-web handed it before.
//
// Unary calls stay on gRPC-web: they are short-lived and never hold a
// connection open.

type Decoder<T> = { fromBinary(bytes: Uint8Array): T }

interface Subscription {
  push: (value: unknown) => void
  fail: (err: Error) => void
  finish: () => void
}

const base64ToBytes = (b64: string): Uint8Array => {
  const bin = atob(b64)
  const out = new Uint8Array(bin.length)
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i)
  return out
}

export class MuxTransport {
  private url: string
  private socket: WebSocket | null = null
  private opening: Promise<WebSocket> | null = null
  private subs = new Map<string, Subscription>()
  private nextId = 1

  constructor(origin: string) {
    // Same origin as the SPA; ws for http, wss for https.
    this.url = origin.replace(/^http/, 'ws') + '/ws'
  }

  private async connect(): Promise<WebSocket> {
    if (this.socket && this.socket.readyState === WebSocket.OPEN) return this.socket
    if (this.opening) return this.opening

    this.opening = new Promise<WebSocket>((resolve, reject) => {
      const ws = new WebSocket(this.url)
      ws.onopen = () => {
        this.socket = ws
        this.opening = null
        resolve(ws)
      }
      ws.onerror = () => {
        this.opening = null
        reject(new Error('mux socket failed to open'))
      }
      ws.onclose = () => {
        this.socket = null
        // Every subscription dies with the socket. Callers already handle a
        // stream ending — the editor's stores reconnect and re-request a full
        // snapshot — so failing them is the honest signal.
        const subs = [...this.subs.values()]
        this.subs.clear()
        for (const s of subs) s.fail(new Error('mux socket closed'))
      }
      ws.onmessage = (ev) => this.dispatch(ev.data)
    })
    return this.opening
  }

  private dispatch(raw: unknown) {
    if (typeof raw !== 'string') return
    let frame: { id?: string; event?: string; error?: string; end?: boolean }
    try {
      frame = JSON.parse(raw)
    } catch {
      return
    }
    if (!frame.id) return
    const sub = this.subs.get(frame.id)
    if (!sub) return

    if (frame.error) {
      this.subs.delete(frame.id)
      sub.fail(new Error(frame.error))
      return
    }
    if (frame.end) {
      this.subs.delete(frame.id)
      sub.finish()
      return
    }
    if (frame.event) sub.push(base64ToBytes(frame.event))
  }

  // stream opens one multiplexed subscription and yields decoded messages,
  // matching the AsyncIterable shape the editor consumes from connect-web —
  // including honouring an AbortSignal.
  stream<T>(kind: string, req: unknown, decoder: Decoder<T>, signal?: AbortSignal): AsyncIterable<T> {
    const id = String(this.nextId++)
    const queue: Uint8Array[] = []
    let waiting: ((v: IteratorResult<Uint8Array>) => void) | null = null
    let failure: Error | null = null
    let done = false

    const sub: Subscription = {
      push: (value) => {
        const bytes = value as Uint8Array
        if (waiting) {
          const w = waiting
          waiting = null
          w({ value: bytes, done: false })
        } else {
          queue.push(bytes)
        }
      },
      fail: (err) => {
        failure = err
        if (waiting) {
          const w = waiting
          waiting = null
          w({ value: undefined as never, done: true })
        }
      },
      finish: () => {
        done = true
        if (waiting) {
          const w = waiting
          waiting = null
          w({ value: undefined as never, done: true })
        }
      }
    }
    this.subs.set(id, sub)

    const cancel = () => {
      this.subs.delete(id)
      const ws = this.socket
      if (ws && ws.readyState === WebSocket.OPEN) {
        ws.send(JSON.stringify({ id, cancel: true }))
      }
      sub.finish()
    }
    signal?.addEventListener('abort', cancel, { once: true })

    const self = this
    return {
      async *[Symbol.asyncIterator]() {
        try {
          // An already-aborted signal never fires 'abort', so the listener
          // below would never clean up: bail before subscribing.
          if (signal?.aborted) return
          const ws = await self.connect()
          ws.send(JSON.stringify({ id, kind, req }))
          for (;;) {
            if (failure) throw failure
            if (queue.length > 0) {
              yield decoder.fromBinary(queue.shift() as Uint8Array)
              continue
            }
            if (done) return
            const next = await new Promise<IteratorResult<Uint8Array>>((resolve) => {
              waiting = resolve
            })
            if (failure) throw failure
            if (next.done) return
            yield decoder.fromBinary(next.value)
          }
        } finally {
          signal?.removeEventListener('abort', cancel)
          self.subs.delete(id)
          const open = self.socket
          if (open && open.readyState === WebSocket.OPEN && !done) {
            open.send(JSON.stringify({ id, cancel: true }))
          }
        }
      }
    }
  }
}

File: internal/flow/mux.go
package flow

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
	"github.com/rs/zerolog/log"
	platform "github.com/tiny-systems/platform-go"
	mcpv1 "github.com/tiny-systems/platform-go/mcp/v1"
	"google.golang.org/grpc"
	"google.golang.org/protobuf/proto"
)

// One socket for every live stream.
//
// The editor keeps three server-streams open per page — the project (or
// flow) stream, statistics, and the activity feed — and gRPC-web runs each
// on its own HTTP request. Served over HTTP/1.1, which is what a browser
// gets on plain http, that costs three of the SIX connections a browser
// allows PER HOST across every tab. Two open tabs exhausted the pool
// exactly, and every request afterwards — including the one carrying widget
// data — queued forever: the dashboard rendered empty cards and the page
// looked frozen.
//
// A WebSocket is one connection and is not subject to that cap, so all the
// streams ride it together, addressed by a caller-chosen id. Events are the
// same protobuf messages the gRPC handlers already produce, base64-framed in
// JSON so a text frame carries them without a bespoke binary format; the
// browser decodes each with the generated message class and receives exactly
// what gRPC-web would have handed it.
//
// This is a tiny-side transport, not a protocol change: the same service
// methods serve both paths, and the hosted platform is untouched.

// muxUpgrader accepts only same-origin upgrades; tiny binds to localhost and
// serves one user, but a browser will happily connect a page from anywhere
// to a local socket, so the check stays.
var muxUpgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool {
		origin := r.Header.Get("Origin")
		if origin == "" {
			return true // non-browser client (tests, curl)
		}
		return origin == "http://"+r.Host || origin == "https://"+r.Host
	},
}

// muxRequest is what the browser sends: open a stream under `id`, or cancel
// one already open.
type muxRequest struct {
	ID     string          `json:"id"`
	Kind   string          `json:"kind"`
	Req    json.RawMessage `json:"req"`
	Cancel bool            `json:"cancel"`
}

// muxFrame is what comes back. Exactly one of Event / Error / End is set.
type muxFrame struct {
	ID    string `json:"id"`
	Event string `json:"event,omitempty"` // base64 protobuf
	Error string `json:"error,omitempty"`
	End   bool   `json:"end,omitempty"`
}

// muxConn serialises writes: gorilla allows one writer at a time and several
// streams share this socket.
type muxConn struct {
	mu   sync.Mutex
	conn *websocket.Conn
}

// muxWriteWait bounds ONE frame write. Writes are serialised on this socket, so
// a browser that stops reading — a suspended tab, a machine asleep — would
// otherwise block the sending goroutine forever while holding the lock, and
// every other stream sharing the socket stalls behind it. The deadline turns
// that into an error the stream can end on. (http.Server's WriteTimeout does
// not apply here: net/http clears the connection's deadlines when the
// WebSocket upgrade hijacks it.)
const muxWriteWait = 30 * time.Second

func (c *muxConn) send(f muxFrame) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if err := c.conn.SetWriteDeadline(time.Now().Add(muxWriteWait)); err != nil {
		return err
	}
	return c.conn.WriteJSON(f)
}

// muxStream adapts a multiplexed subscription to the grpc.ServerStreamingServer
// interface the existing handlers expect. Only Send and Context are ever
// called on a server stream; the embedded interface satisfies the rest of the
// signature without pretending to implement it.
type muxStream[T proto.Message] struct {
	grpc.ServerStream
	ctx  context.Context
	id   string
	conn *muxConn
}

func (s *muxStream[T]) Context() context.Context { return s.ctx }

func (s *muxStream[T]) Send(msg T) error {
	raw, err := proto.Marshal(msg)
	if err != nil {
		return err
	}
	return s.conn.send(muxFrame{ID: s.id, Event: base64.StdEncoding.EncodeToString(raw)})
}

// muxServices is the set of streaming backends the socket can address — the
// same values registered on the gRPC server, so both transports serve
// identical handlers.
type muxServices struct {
	flow     *Service
	project  projectService
	stats    statisticsService
	activity workspaceActivityService
}

// handler serves the socket. Each subscription runs in its own goroutine so a
// slow stream never blocks its neighbours.
func (s muxServices) handler(w http.ResponseWriter, r *http.Request) {
	conn, err := muxUpgrader.Upgrade(w, r, nil)
	if err != nil {
		return // Upgrade already answered the request
	}
	defer func() { _ = conn.Close() }()

	mc := &muxConn{conn: conn}
	ctx, cancelAll := context.WithCancel(r.Context())
	defer cancelAll()

	var mu sync.Mutex
	cancels := map[string]context.CancelFunc{}

	for {
		var req muxRequest
		if err := conn.ReadJSON(&req); err != nil {
			return // client went away
		}
		if req.ID == "" {
			continue
		}

		if req.Cancel {
			mu.Lock()
			if cancel, ok := cancels[req.ID]; ok {
				cancel()
				delete(cancels, req.ID)
			}
			mu.Unlock()
			continue
		}

		streamCtx, cancel := context.WithCancel(ctx)
		mu.Lock()
		if prev, ok := cancels[req.ID]; ok {
			prev() // a re-subscribe on the same id replaces the old stream
		}
		cancels[req.ID] = cancel
		mu.Unlock()

		go func(req muxRequest, streamCtx context.Context) {
			err := s.serve(streamCtx, mc, req)
			mu.Lock()
			delete(cancels, req.ID)
			mu.Unlock()
			if streamCtx.Err() != nil {
				return // cancelled by the client; it isn't waiting for a reply
			}
			if err != nil {
				_ = mc.send(muxFrame{ID: req.ID, Error: err.Error()})
				return
			}
			_ = mc.send(muxFrame{ID: req.ID, End: true})
		}(req, streamCtx)
	}
}

// serve dispatches one subscription to the service method that backs it. The
// handlers are the same ones gRPC-web calls — this only changes how their
// messages reach the browser.
func (s muxServices) serve(ctx context.Context, mc *muxConn, req muxRequest) error {
	switch req.Kind {
	case "project.getStream":
		var in platform.GetProjectStreamRequest
		if err := unmarshalMuxReq(req.Req, &in); err != nil {
			return err
		}
		return s.project.GetStream(&in, &muxStream[*platform.GetProjectStreamEvent]{ctx: ctx, id: req.ID, conn: mc})

	case "flow.getFlowStream":
		var in platform.GetFlowStreamRequest
		if err := unmarshalMuxReq(req.Req, &in); err != nil {
			return err
		}
		return s.flow.GetFlowStream(&in, &muxStream[*platform.GetFlowStreamResponse]{ctx: ctx, id: req.ID, conn: mc})

	case "statistics.getStream":
		var in platform.StatisticsStreamRequest
		if err := unmarshalMuxReq(req.Req, &in); err != nil {
			return err
		}
		return s.stats.GetStream(&in, &muxStream[*platform.StatisticsStreamResponse]{ctx: ctx, id: req.ID, conn: mc})

	case "activity.watch":
		var in mcpv1.WatchWorkspaceActivityRequest
		if err := unmarshalMuxReq(req.Req, &in); err != nil {
			return err
		}
		return s.activity.Watch(&in, &muxStream[*mcpv1.WorkspaceActivityEvent]{ctx: ctx, id: req.ID, conn: mc})
	}

	log.Debug().Str("kind", req.Kind).Msg("mux: unknown stream kind")
	return errUnknownMuxKind(req.Kind)
}

// unmarshalMuxReq accepts the browser's JSON request. Field names come from
// the generated types, so the same object the gRPC-web client would send
// decodes here unchanged.
func unmarshalMuxReq(raw json.RawMessage, out interface{}) error {
	if len(raw) == 0 {
		return nil
	}
	return json.Unmarshal(raw, out)
}

type muxKindError string

func (e muxKindError) Error() string { return "unknown stream kind: " + string(e) }

func errUnknownMuxKind(kind string) error { return muxKindError(kind) }
